File: backend/app/services/freshness_engine.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Literal


SPECIES_SENSITIVITY = {
    "prawns": 1.25,
    "sardine": 1.2,
    "pomfret": 1.1,
    "seer fish": 0.9,
}


def _hours_between(start: datetime, end: datetime) -> float:
    return max((end - start).total_seconds() / 3600, 0.0)


def _label(score: float) -> Literal["Excellent", "High", "Moderate", "At Risk"]:
    if score >= 90:
        return "Excellent"
    if score >= 80:
        return "High"
    if score >= 70:
        return "Moderate"
    return "At Risk"


def _spoilage_risk(score: float) -> Literal["Low", "Medium", "High"]:
    if score >= 85:
        return "Low"
    if score >= 70:
        return "Medium"
    return "High"


def calculate_freshness_intelligence(
    *,
    catch_time: datetime,
    landing_time: datetime,
    packing_time: datetime,
    dispatch_time: datetime,
    arrival_eta: datetime,
    species: str,
    cold_chain_ok: bool,
    reference_time: datetime | None = None,
) -> dict:
    """Return freshness intelligence and pricing hints using weighted multi-factor logic."""
    now = reference_time or datetime.now(timezone.utc)
    species_factor = SPECIES_SENSITIVITY.get(species.lower(), 1.0)

    hours_since_catch = _hours_between(catch_time, now)
    packing_delay = _hours_between(landing_time, packing_time)
    dispatch_delay = _hours_between(packing_time, dispatch_time)
    eta_hours = _hours_between(now, arrival_eta)

    score = 100.0
    audit = []

    # 1. Catch to Now (Overall Age)
    age_penalty = round(hours_since_catch * 1.2 * species_factor, 1)
    score -= age_penalty
    audit.append({
        "label": "Time Since Catch",
        "value": f"{round(hours_since_catch, 1)}h",
        "status": "pass" if hours_since_catch < 12 else "warn",
        "impact": f"-{age_penalty} pts"
    })

    # 2. Thermal Stability
    if cold_chain_ok:
        thermal_bonus = 2.5
        score += thermal_bonus
        audit.append({
            "label": "Thermal Stability",
            "value": "2°C - 4°C",
            "status": "pass",
            "impact": f"+{thermal_bonus} pts (Stable)"
        })
    else:
        score -= 15
        audit.append({
            "label": "Cold-Chain Integrity",
            "value": "Drift Detected",
            "status": "at_risk",
            "impact": "-15.0 pts"
        })

    # 3. Processing Efficiency
    process_penalty = round(packing_delay * 0.8 * species_factor, 1)
    score -= process_penalty
    audit.append({
        "label": "Hub Processing",
        "value": f"Landed to Packed in {round(packing_delay, 1)}h",
        "status": "pass" if packing_delay < 3 else "info",
        "impact": f"-{process_penalty} pts"
    })

    # 4. Species Sensitivity
    if species_factor > 1.1:
        audit.append({
            "label": "Species Stability",
            "value": f"{species.capitalize()} (High Decay)",
            "status": "info",
            "impact": "Increased weighting applied"
        })

    score = max(0.0, min(100.0, round(score, 1)))
    return {
        "freshness_score": score,
        "freshness_label": _label(score),
        "spoilage_risk": _spoilage_risk(score),
        "flash_drop": score < 80,
        "audit": audit
    }

File: src/ml/freshness_engine.ts
/**
 * Freshness Intelligence + Zero-Waste Pricing Engine
 * 
 * Simulated ML/Rules Engine:
 * Takes temporal variables (catch, landing, packing, dispatch) and computes
 * an exact freshness confidence score (0-100) and executes dynamic pricing 
 * thresholds to enforce zero-waste unit economics.
 */

export interface LogisticsInput {
  catchTime: Date;
  landingTime: Date;
  packingTime: Date;
  dispatchTime: Date;
  arrivalEta: Date;
  species: 'Prawns' | 'Sardine' | 'Seer Fish' | 'Pomfret';
  coldChainMaintained: boolean;
  basePricePerKg: number;
}

export interface IntelligenceOutput {
  freshnessScore: number;
  freshnessLabel: 'Excellent' | 'High' | 'Moderate' | 'At Risk';
  dynamicPrice: number;
  flashDrop: boolean;
  spoilageRisk: 'Low' | 'Medium' | 'High' | 'Critical';
  wastagePercentage: number;
}

const SPECIES_SENSITIVITY: Record<string, number> = {
  'Prawns': 1.8,    // Highly sensitive
  'Sardine': 1.6,   // Highly sensitive
  'Pomfret': 1.2,   // Medium
  'Seer Fish': 1.0  // More stable
};

/**
 * Computes difference in hours between two dates.
 */
const diffHours = (start: Date, end: Date): number => {
  return Math.max(0, (end.getTime() - start.getTime()) / (1000 * 60 * 60));
};

export const calculateFreshnessIntelligence = (input: LogisticsInput): IntelligenceOutput => {
  let score = 100;
  
  // 1. Calculate time gaps
  const hoursSinceCatch = diffHours(input.catchTime, new Date());
  const packingDelay = diffHours(input.landingTime, input.packingTime);
  const dispatchDelay = diffHours(input.packingTime, input.dispatchTime);
  const transitTime = diffHours(input.dispatchTime, new Date());
  
  // 2. Identify Species Multiplier
  const sensitivityMultiplier = SPECIES_SENSITIVITY[input.species] || 1.5;

  // 3. Apply Time Degradation Penalties
  // Baseline decay simply for being out of ocean
  score -= (hoursSinceCatch * 1.5 * sensitivityMultiplier);
  
  // Severe penalty for delays arriving to the cold chain (landing to packing)
  score -= (packingDelay * 2.5 * sensitivityMultiplier);
  
  // Mild penalty for dispatch idle time
  score -= (dispatchDelay * 1.0);
  
  // Transit degradation
  score -= (transitTime * 0.8 * sensitivityMultiplier);

  // 4. Cold Chain Break Penalty (Absolute Dealbreaker)
  if (!input.coldChainMaintained) {
    score -= 35; // Massive penalty for breaking temp limits
  }

  // Bound score
  score = Math.max(0, Math.min(100, Math.round(score)));

  // 5. Determine State Labels
  let freshnessLabel: IntelligenceOutput['freshnessLabel'];
  let spoilageRisk: IntelligenceOutput['spoilageRisk'];

  if (score >= 90) {
    freshnessLabel = 'Excellent';
    spoilageRisk = 'Low';
  } else if (score >= 80) {
    freshnessLabel = 'High';
    spoilageRisk = 'Medium';
  } else if (score >= 70) {
    freshnessLabel = 'Moderate';
    spoilageRisk = 'High';
  } else {
    freshnessLabel = 'At Risk';
    spoilageRisk = 'Critical';
  }

  // 6. Dynamic Pricing (Zero-Waste Logic)
  // Ensure unit economics are preserved while minimizing raw spoilage loss.
  let dynamicPrice = input.basePricePerKg;
  let flashDrop = false;

  if (score >= 90) {
    // 90+ -> Full Premium Price
    dynamicPrice = input.basePricePerKg;
  } else if (score >= 80) {
    // 80-89 -> Mild markdown (5-8%)
    dynamicPrice = Math.round(input.basePricePerKg * 0.94);
  } else if (score >= 70) {
    // 70-79 -> Flash Drop to clear stock fast (15-20% off)
    dynamicPrice = Math.round(input.basePricePerKg * 0.82);
    flashDrop = true;
  } else {
    // <70 -> Aggressive markdown just to recover logistics cost (40% off)
    dynamicPrice = Math.round(input.basePricePerKg * 0.60);
    flashDrop = true;
  }

  // 7. Wastage Probability Proxy (%)
  let wastagePercentage = 0;
  if (score >= 90) {
    wastagePercentage = 4;
  } else if (score >= 80) {
    wastagePercentage = 9;
  } else if (score >= 70) {
    wastagePercentage = 18;
  } else {
    wastagePercentage = 32;
  }

  if (!input.coldChainMaintained) {
    wastagePercentage += 10;
  }
  wastagePercentage = Math.max(0, Math.min(95, Math.round(wastagePercentage)));

  return {
    freshnessScore: score,
    freshnessLabel,
    dynamicPrice,
    flashDrop,
    spoilageRisk,
    wastagePercentage,
  };
};
